core/src/main/java/org/apache/iceberg/avro/PruneColumns.java

```diff
@@ -138,28 +166,98 @@ public Schema array(Schema array, Schema element) {
 
   @Override
   public Schema map(Schema map, Schema value) {
-    int keyId = AvroSchemaUtil.getKeyId(map);
-    int valueId = AvroSchemaUtil.getValueId(map);
+    Integer keyId = id(map, AvroSchemaUtil.KEY_ID_PROP, "key");
+    Integer valueId = id(map, AvroSchemaUtil.VALUE_ID_PROP, "value");
+    if (keyId == null) {
+      if (valueId != null) {
+        LOG.warn("Map schema {} has value-id but not key-id", map);
+      }
+      return null;
+    }
     // if either key or value is selected, the whole map must be projected
     if (selectedIds.contains(keyId) || selectedIds.contains(valueId)) {
-      return map;
+      // Assign ids. Ids may not always be present in the schema
+      return mapWithIds(map, keyId, valueId);
     } else if (value != null) {
       if (value != map.getValueType()) {
         // the value must be a projection
-        return Schema.createMap(value);
+        return mapWithIds(Schema.createMap(value), keyId, valueId);
       }
       return map;
     }
 
     return null;
   }
 
+  private Schema arrayWithId(Schema array, Integer elementId) {
+    if (!AvroSchemaUtil.hasProperty(array, AvroSchemaUtil.ELEMENT_ID_PROP)) {
+      Schema result = Schema.createArray(array.getElementType());
+      result.addProp(AvroSchemaUtil.ELEMENT_ID_PROP, elementId);
+      return result;
+    }
+    return array;
+  }
+
+  private Schema complexMapWithIds(Schema map, Integer keyId, Integer valueId) {
+    Schema keyValue = map.getElementType();
+    if (!AvroSchemaUtil.hasFieldId(keyValue.getField("key"))) {
+      return AvroSchemaUtil.createMap(
+          keyId, keyValue.getField("key").schema(),
+          valueId, keyValue.getField("value").schema());
+    }
+    return map;
+  }
+
+  private Schema mapWithIds(Schema map, Integer keyId, Integer valueId) {
+    if (!AvroSchemaUtil.hasProperty(map, AvroSchemaUtil.KEY_ID_PROP)) {
+      Schema result = Schema.createMap(map.getValueType());
+      result.addProp(AvroSchemaUtil.KEY_ID_PROP, keyId);
+      result.addProp(AvroSchemaUtil.VALUE_ID_PROP, valueId);
+      return result;
+    }
+    return map;
+  }
+
   @Override
   public Schema primitive(Schema primitive) {
     // primitives are not selected directly
     return null;
   }
 
+  private Integer id(Schema schema, String propertyName, String mappedName) {
+    if (AvroSchemaUtil.hasProperty(schema, propertyName)) {
+      return AvroSchemaUtil.getId(schema, propertyName);
+    } else {
+      MappedField mappedField = mappedField(mappedName);
+      if (mappedField != null) {
+        return mappedField.id();
+      } else {
+        return null;
+      }
+    }
+  }
+
+  private Integer fieldId(Schema.Field field) {
```

Comment:
What about updating `AvroSchemaUtil.getFieldId` to accept a mapping? Is that used in more places where you wouldn't want to add a mapping? Seems like the changes might be simpler if the mapping was supported in util.

Reply:
Today, `AvroSchemaUtil.getFieldId` is used by

1. BuildAvroProjection
2. SchemaToType
3. PruneColumns
4. SparkAvroWriter

Except 3, others will not use NameMapping. Also I will also have to pass a list of `fieldNames` together with the mapping which complicates it a little.

Let me know if you still feel that we should make this change. I'll update it then

Reply:
I ended up incorporating your suggestion